core/Tracker/Db/Pdo/Mysql.php

```diff
@@ -209,6 +218,57 @@ public function fetch($query, $parameters = array())
      * @throws DbException if an exception occurred
      */
     public function query($query, $parameters = array())
+    {
+        try {
+            return $this->executeQuery($query, $parameters);
+        } catch (Exception $e) {
+            $isSelectQuery = stripos(trim($query), 'select ') === 0;
+            $isUpdateQuery = stripos(trim($query), 'update ') === 0;
+
+            if (($isSelectQuery || $isUpdateQuery)
+                && !$this->activeTransaction
+                && $this->isMysqlServerHasGoneAwayError($e)) {
+                // mysql may return a MySQL server has gone away error when trying to execute the query
+                // in that case we want to retry establishing the connection once after a short sleep
+                // we're only retrying SELECT and UPDATE queries to prevent inserting records twice for some reason
+                // when transactions are used, then we just want it to fail as we'd be only writing partial data
+                $this->reconnect($e);
+                return $this->executeQuery($query, $parameters);
+            } else {
+                $message = $e->getMessage() . " In query: $query Parameters: " . var_export($parameters, true);
+                throw new DbException("Error query: " . $message, (int) $e->getCode());
+            }
+
+        }
+    }
+
+    /**
+     * @internal for tests only
+     * @param Exception $e
+     * @throws Exception
+     */
+    public function reconnect(Exception $e)
+    {
+        $this->disconnect();
+        usleep(100 * 1000); // wait for 100ms
```

Comment:
In tracker we're now using 100ms which we've been testing on production and worked well enough for us